macros/src/main/scala/scala/tools/abide/traversal/WarningsMacro.scala

```diff
@@ -0,0 +1,73 @@
+package scala.tools.abide.traversal
+
+import scala.reflect.macros._
+import scala.language.experimental.macros
+
+/**
+ * WarningExtractor
+ *
+ * Provide the extraction macro that uses the placeholder [[WarningExtractor.tree]] value in the
+ * [[WarningExtractor]] class with to create a function from Tree to String by replacing the
+ * dummy tree with the function parameter.
+ *
+ * To get the String we use the StringContext that we assume is the macro caller and
+ * apply the `StringContext.s` method to the transformed arguments. We also assume all
+ * references to `_.this.tree` refer to the [[WarningExtractor.tree]] value as checking for
+ * enclosing context is non-trivial and discouraged.
+ */
+object WarningExtractor {
```

Comment:
It's neat, and I like how much more concise rules get! But I'm wondering if this isn't too much magic for an essentially very simple logic. If I'm right, the string interpolator and macro allow one to write

```scala
val warning = w"Illegal $tree"
//...inside traversal
nok(tree)
```
instead of 
```scala
nok(new Warning(tree, s"Illegal $tree"))
```

In the current implementation I'm not sure it's possible to use the interpolator in any other context than top-level (so that `tree` binds to `this.tree`, as expected by the macro implementation. For example, I'd find it surprising if this didn't work:
```scala
nok(w"Illegal ${t.children(0)}")
```

What do you think?